### PR TITLE
Initial State Machine Implementation

### LINKED ISSUE #10: Implement State Machine
Implement a Deterministic Finite State Machine.

### PR DESCRIPTION
**Pull request by Byrdman32**
Initial work for implementing a state machine into Autonomy. Future Pull Requests will implement the processes that happen within each state and implement the RoveComm handlers that will have the ability to start and stop the state machine.

Associated with #10

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -92,6 +92,7 @@ file(GLOB Tools_SRC			 	CONFIGURE_DEPENDS  "tools/*.cpp")
 add_executable(${PROJECT_NAME}   ${External_SRC}
 								 ${Util_SRC}
 								 ${Algorithms_SRC}
+                                 ${Drivers_SRC}
```
Comment (ClayJay3) [inline CMakeLists.txt]: Drivers_SRC is provided twice.

File: src/AutonomyConstants.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ * @author Eli Byrd (edbgkk@mst.edu)
+ * @date 2023-07-30
+ *
+ * @copyright Copyright MRDT 2023 - All Rights Reserved
+ ******************************************************************************/
+
+#ifndef CONSTS_H
+#define CONSTS_H
+
+// Drive Constants
```
Comment (ClayJay3) [inline src/AutonomyConstants.h]: This should be in a namespace so we can do typedefs.

namespace constants
{
    // Drive constants.
    const int MAX_DRIVE_POWER = 250;
    const int MIN_DRIVE_POWER = 50;
}    // namespace constants

File: src/interfaces/StateMachine.hpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+ *
+ *
+ * @author Eli Byrd (edbgkk@mst.edu)
+ * @date 2023-07-17
+ ******************************************************************************/
+struct Idle_AbortTransition;
+struct Idle_NavigatingTransition;
+struct Idle_ReverseTransition;
+
+#include "../states/IdleState.hpp"
```
Comment (ClayJay3) [inline src/interfaces/StateMachine.hpp:54]: Why are the state transitions redeclared here and then defined when included? All #includes should be at the top of the file and then struct objects should be declared. There might be a more elegant way to do these state transitions.

**Review by ClayJay3 — COMMENTED** [review]
All state files need doxygen comment blocks for every struct.

File: src/interfaces/StateMachine.hpp (hunk 3 above)
Comment (Byrdman32) [reply src/interfaces/StateMachine.hpp:54]: > Why are the state transitions redeclared here and then defined when included? All #includes should be at the top of the file and then struct objects should be declared. There might be a more elegant way to do these state transitions.

The states have to be declared after the state machine object is declared. This was the cleanest way I have come up with but am open to suggestions.

**Review by ClayJay3 — CHANGES_REQUESTED** [review]
All structs should have simple doxygen comment blocks.

**Review by ClayJay3 — APPROVED** [review]

**Pull request merged**